HeterogeneousCore/CUDAServices/plugins/NVProfilerService.cc

```diff
@@ -715,351 +796,409 @@ NVProfilerService::preStreamEndLumi(edm::StreamContext const& sc) {
 void
 NVProfilerService::postStreamEndLumi(edm::StreamContext const& sc) {
   auto sid = sc.streamID();
-  nvtxDomainRangePop(stream_domain(sid));
+  if (not skipFirstEvent_ or streamFirstEventDone_[sid]) {
+    nvtxDomainRangePop(stream_domain(sid));
+  }
 }
 
 void
 NVProfilerService::preEvent(edm::StreamContext const& sc) {
   auto sid = sc.streamID();
-  event_[sid] = nvtxDomainRangeStartColor(stream_domain(sid), "event", nvtxDarkGreen);
+  if (not skipFirstEvent_ or streamFirstEventDone_[sid]) {
+    event_[sid] = nvtxDomainRangeStartColor(stream_domain(sid), "event", nvtxDarkGreen);
+  }
 }
 
 void
 NVProfilerService::postEvent(edm::StreamContext const& sc) {
   auto sid = sc.streamID();
-  nvtxDomainRangeEnd(stream_domain(sid), event_[sid]);
+  if (not skipFirstEvent_ or streamFirstEventDone_[sid]) {
+    nvtxDomainRangeEnd(stream_domain(sid), event_[sid]);
+    event_[sid] = nvtxInvalidRangeId;
+  } else {
+    streamFirstEventDone_[sid] = true;
+    // there is a possible race condition among different threads processing different events;
+    // however, cudaProfilerStart() is supposed to be thread-safe and ignore multiple calls, so this should not be an issue.
+    if (std::all_of(streamFirstEventDone_.begin(), streamFirstEventDone_.end(), [](bool x){ return x; })) {
+      globalFirstEventDone_ = true;
```

Comment:
We discussed a bit with @Dr15Jones, and we think that `streamFirstEventDone_` and `globalFirstEventDone_` should be changed to use `std::atomic<bool>`. Especially the `vector<bool>` is dangerous because it is internally a bit pattern, so setting one bit in current thread may invalidate an operation on a different bit on a same word in another thread.

Reply:
@makortel @Dr15Jones thanks for the suggestion.
Do the changes to the member variables (lines 296-297 ) and to the initialisation (lines 541-543) look correct ?

Reply:
Looks good to me, thanks. You could also use `globalFirstEventDone_.compare_exchange_strong()` here to call `cudaProfilerStart()` exactly once (that would also allow to remove the comments above).

Reply:
Thanks for the suggestion.
I have also changed a bit the way in which I initialise the vector, as this "feels" better - though I expect the end result to be the same.